fud/fud/stages/verilator/numeric_types.py

```diff
@@ -0,0 +1,296 @@
+    is_signed: bool
+    string_repr: str
+    bit_string_repr: str
+    hex_string_repr: str
+    uint_repr: int
+
+    def __init__(self, value: str, width: int, is_signed: bool):
+        if not isinstance(value, str) or len(value) == 0:
+            raise InvalidNumericType(
+                f"The value: {value} of type: "
+                f"{type(value)} should be a non-empty string."
+            )
+        if value.startswith("-") and not is_signed:
+            raise InvalidNumericType(
+                f"A negative value was provided: {value}, " f"and `is_signed` is False."
+            )
+        value = value.strip()
+        self.width = width
+        self.is_signed = is_signed
+
+        if value.startswith("0b"):
+            self.string_repr = str(int(value, 2))
+            # Zero padding for bit string.
+            self.bit_string_repr = "0" * (width - len(value)) + value[2:]
+
+            self.uint_repr = int(self.bit_string_repr, 2)
+            self.hex_string_repr = np.base_repr(self.uint_repr, 16)
+        elif value.startswith("0x"):
+            self.string_repr = str(int(value, 16))
+            self.hex_string_repr = value[2:]
+            self.bit_string_repr = np.binary_repr(
+                int(self.hex_string_repr, 16), self.width
+            )
+            self.uint_repr = int(self.bit_string_repr, 2)
+        else:
+            self.string_repr = value
+
+    def str_value(self) -> str:
+        return self.string_repr
+
+    def bit_string(self, with_prefix=True) -> str:
+        return f"{'0b' if with_prefix else ''}{self.bit_string_repr}"
+
+    def hex_string(self, with_prefix=True) -> str:
+        return f"{'0x' if with_prefix else ''}{self.hex_string_repr}"
+
+    def unsigned_integer(self) -> int:
+        return self.uint_repr
+
+    def pretty_print(self):
+        pass
+
+
+@dataclass
+class Bitnum(NumericType):
+    """Represents a two's complement bitnum."""
+
+    def __init__(self, value: str, width: int, is_signed: bool):
+        super().__init__(value, width, is_signed)
+        integer_value = int(self.string_repr)
+
+        if all(x not in value for x in ["0x", "0b"]):
+            # The actual value was passed instead of a base string representation.
+            self.bit_string_repr = np.binary_repr(integer_value, self.width)
+            self.uint_repr = int(self.bit_string_repr, 2)
+            self.hex_string_repr = np.base_repr(self.uint_repr, 16)
+
+        if is_signed and self.uint_repr > (2 ** (width - 1)):
+            self.string_repr = str(-1 * ((2 ** width) - self.uint_repr))
+
+        if len(self.bit_string_repr) > width:
+            raise InvalidNumericType(
+                f"The value: {value} will overflow when trying to represent"
+                f"{len(self.bit_string_repr)} bits with width: {width}"
+            )
+
+    def pretty_print(self):
+        print(
+            f"""
+            {'Signed' if self.is_signed else ''} Bitnum: {self.string_repr}
```

Comment:
There is whitespace in this string at the start of each line. Maybe you want the string completely left indented?